### PR TITLE
Design tools unable create DbContext with custom constructor using the services provider

### PR DESCRIPTION
**Pull request by VladDragnea**
I've noticed that design time tools cannot instantiate a DbContext if the constructors doesn't follow a standard pattern (default constructor or with DbContextOptions derived parameter), even if the service provider is able to create the context.
The proposed change only replace:
```
Activator.CreateInstance(...);
```
with:
```
ActivatorUtilities.GetServiceOrCreateInstance(...);
```
which I think it will fallback to the previous behavior if the service provider is unable to create the context.
A test for checking this scenario has been added.

### CODE REVIEW

**Comment by bricelam** [comment]
This will only help when your DbContext isn't registered as a service, but it has a constructor that takes services. Is that the case you're trying to fix?

**Comment by VladDragnea** [comment]
Yes, but not only that. I could also happen when my DbContext is registered as a service, but not via AddDbContext. For example a context with a single constructor that takes a connection string, registered as a service with a factory. It will be provided by the service provider but not discoverable, because the code that looks for it actually looks for DbContextOptions.

**Comment by ajcvickers** [comment]
@VladDragnea If you have a single constructor taking the connection string, then how will the context created using `ActivatorUtilities` get the correct connection string?

**Comment by VladDragnea** [comment]
@ajcvickers the parameters, whichever they are, could be provided by the factory that is registered with the service provider. I've simulated this situation in the test case with a dummy connection string.

File: test/EFCore.Design.Tests/Design/Internal/DbContextOperationsCustomTest.cs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        }
+
+        private static class TestProgram
+        {
+#pragma warning disable RCS1213 // Remove unused member declaration.
+            private static TestWebHost BuildWebHost(string[] args)
+#pragma warning restore RCS1213 // Remove unused member declaration.
+                => new TestWebHost(
+                    new ServiceCollection()
+                    .AddTransient((_) => new TestContextWithCustomConstructor(nameof(DbContextActivatorTest)))
```
Comment (bricelam) [inline test/EFCore.Design.Tests/Design/Internal/DbContextOperationsCustomTest.cs]: Won't this test pass without your change? If I remember correctly, we try asking for the context from the service provider long before resorting to Activator.CreateInstance()

Comment (bricelam) [reply test/EFCore.Design.Tests/Design/Internal/DbContextOperationsCustomTest.cs]: The the test fails.

**Comment by bricelam** [comment]
Made a few tweaks:
* Handled the different exception type thrown by ActivatorUtilities
* Merged your test into our existing DbContextOperations tests and changed it into a more real-world scenario

**Pull request merged**